File: src/csharp/Algorithms/CotanLaplacian.cs
using System;
using System.Collections.Generic;
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;
using DDGCompanion.Core;

namespace DDGCompanion.Algorithms
{
    public class CotanLaplacian
    {
        public static SparseMatrix Build(Mesh mesh)
        {
            int n = mesh.Vertices.Count;
            var builder = new SparseMatrix(n, n);
            
            foreach (var v in mesh.Vertices)
            {
                int i = v.Index;
                double sumWeights = 0;
                var he = v.HalfEdge;
                if (he == null)
                {
                    continue;
                }

                var visited = new HashSet<int>();
                while (he != null && visited.Add(he.Index))
                {
                    if (he.Vertex == null)
                    {
                        break;
                    }

                    int j = he.Vertex.Index;
                    double weight = ComputeCotanWeight(he);
                    builder[i, j] += weight;
                    sumWeights += weight;

                    if (he.Twin?.Next == null) break;
                    he = he.Twin.Next;
                    if (he == v.HalfEdge) break;
                }
                builder[i, i] = -sumWeights;
            }
            return builder;
        }
        
        public static SparseMatrix BuildMassMatrix(Mesh mesh)
        {
            int n = mesh.Vertices.Count;
            var builder = new SparseMatrix(n, n);
            
            // Barycentric dual area for each vertex
            foreach (var vertex in mesh.Vertices)
            {
                double area = 0.0;
                
                if (vertex.HalfEdge != null)
                {
                    var he = vertex.HalfEdge;
                    do
                    {
                        if (he.Face != null)
                        {
                            // Add 1/3 of the triangle area
                            area += he.Face.Area() / 3.0;
                        }
                        he = he.Twin?.Next;
                    } while (he != null && he != vertex.HalfEdge);
                }
                
                builder[vertex.Index, vertex.Index] = area;
            }
            
            return builder;
        }
        
        private static double ComputeCotanWeight(HalfEdge he)
        {
            if (he.Edge == null) return 0.0;
            var cotan = he.Edge.Cotan();
            return double.IsFinite(cotan) ? cotan : 0.0;
        }
    }
}

File: src/cpp/algorithms/CotanLaplacian.cpp
#include "CotanLaplacian.h"
#include <Eigen/SparseLU>
#include <cmath>

Eigen::SparseMatrix<double> CotanLaplacian::build(const Mesh& mesh) {
    int n = mesh.numVertices();
    Eigen::SparseMatrix<double> L(n, n);
    std::vector<Eigen::Triplet<double>> triplets;
    
    for (const auto& v : mesh.vertices) {
        int i = v->index;
        double sumW = 0;
        HalfEdge* he = v->halfedge;
        do {
            int j = he->next->vertex->index;
            double w = 0;
            if (he->face) {
                auto e0 = he->vector();
                auto e1 = he->next->vector();
                w += 0.5 * (-e0.dot(e1)) / e0.cross(e1).norm();
            }
            if (he->twin->face) {
                auto e0 = he->twin->vector();
                auto e1 = he->twin->next->vector();
                w += 0.5 * (-e0.dot(e1)) / e0.cross(e1).norm();
            }
            triplets.push_back({i, j, w});
            sumW += w;
            he = he->twin->next;
        } while (he != v->halfedge);
        triplets.push_back({i, i, -sumW});
    }
    L.setFromTriplets(triplets.begin(), triplets.end());
    return L;
}

Eigen::SparseMatrix<double> CotanLaplacian::buildMassMatrix(const Mesh& mesh) {
    auto areas = computeVertexAreas(mesh);
    Eigen::SparseMatrix<double> M(mesh.numVertices(), mesh.numVertices());
    std::vector<Eigen::Triplet<double>> triplets;
    for (int i = 0; i < areas.size(); i++)
        triplets.push_back({i, i, areas(i)});
    M.setFromTriplets(triplets.begin(), triplets.end());
    return M;
}

Eigen::VectorXd CotanLaplacian::computeVertexAreas(const Mesh& mesh) {
    Eigen::VectorXd areas = Eigen::VectorXd::Zero(mesh.numVertices());
    for (const auto& f : mesh.faces) {
        double A = f->area() / 3.0;
        for (auto v : f->vertices())
            areas(v->index) += A;
    }
    return areas;
}

Eigen::MatrixXd CotanLaplacian::solvePoisson(const Mesh& mesh, const Eigen::MatrixXd& F) {
    auto L = build(mesh);
    auto M = buildMassMatrix(mesh);
    Eigen::SparseLU<Eigen::SparseMatrix<double>> solver;
    solver.compute(L);
    return solver.solve(M * F);
}
